### ISSUE TITLE
Color bar for scalar visualization

### ISSUE THREAD

**Issue by lslog3** [libigl/libigl#1534]
#### Describe your feature request

Is it possible to have a color bar to display the scalar field numerical values, e.g. Gaussian curvature? viewer.data().set_data() allows to present a scalar field on vertices, but what the numbers these colors present are not clear. This is why a color bar can be useful.

**Comment by jdumas** [comment]
Yes it would be nice to have such a feature. I provided a simple colorbar plugin as an answer to # #1388, but I never made it into a proper plugin.

**Comment by lslog3** [comment]
> Yes it would be nice to have such a feature. I provided a simple colorbar plugin as an answer to # #1388, but I never made it into a proper plugin.

Thank you very much. This is really helpful.

**Comment by lslog3** [comment]
> Yes it would be nice to have such a feature. I provided a simple colorbar plugin as an answer to # #1388, but I never made it into a proper plugin.

I am sorry, just get started with using libigl. I am struggling run the colorbar glugin codes. My codes are like this:

---------------------------------------------------------------------------------
   // Plot the mesh
  viewer.data().set_mesh(V, F);
  viewer.data().set_data(Eigen::VectorXd::Map(&r_[0], r_.size()));

  // Plot Colorbar
  ColorbarPlugin cbar;
  cbar.init_colormaps();
  cbar.draw_colormap_combo();
  auto min_max = std::minmax_element(std::begin(r_), std::end(r_));
  Eigen::Vector4f bgcolor(0, 0, 0.04, 1);
  cbar.draw_colorbar(igl::COLOR_MAP_TYPE_JET, *min_max.first, *min_max.second, bgcolor);

  viewer.launch();
---------------------------------------------------------------------------------

I got a runtime error triggered by glGenTextures(1, &id);

What is wrong with my code? Any idea? Thanks.

**Comment by jdumas** [comment]
You should not be calling `init_colormaps()` at every draw iteration, just once when you setup the viewer and register the plugin.

**Comment by lslog3** [comment]
> You should not be calling `init_colormaps()` at every draw iteration, just once when you setup the viewer and register the plugin.

Thanks. Is there any example code I can follow?

**Comment by jdumas** [comment]
No

**Comment by lslog3** [comment]
> No

Hi I am making some efforts to implement this. Start to understand what you are saying...

I have ColorbarPlugin derived from ViewerPlugin.

#pragma once
#include <igl/colormap.h>
#include <array>

#include <glad/glad.h>
#include <igl/opengl/glfw/ViewerPlugin.h>
#include <igl/opengl/glfw/Viewer.h>


// Forward declarations
struct ImGuiContext;

class ColorbarPlugin: public igl::opengl::glfw::ViewerPlugin {
public:
    // Override parament virtual functions
    void init(igl::opengl::glfw::Viewer* _viewer);
    bool pre_draw();
    bool post_draw();
    void shutdown();

    void init_colormaps();

    int draw_colormap_combo() const;

    void draw_colorbar(igl::ColorMapType cm, float xmin, float xmax,
        const Eigen::Vector4f& background_color) const;

    void set_max_min(float max, float min);

protected:
    std::array<GLuint, igl::NUM_COLOR_MAP_TYPES> colormaps_;

    igl::ColorMapType colormap_type_ = igl::COLOR_MAP_TYPE_VIRIDIS;

    float max_ = 0.0;
    float min_ = 0.0;

    // ImGui Context
    ImGuiContext* context_ = nullptr;
};

Here is some implementation of ColorbarPlugin. Basically copy the same style as ImGuiMenu
#include "ColorbarPlugin.h"
#include <imgui/imgui.h>
#include <imgui_impl_glfw.h>
#include <imgui_impl_opengl3.h>

void ColorbarPlugin::set_max_min(float max, float min)
{
    max_ = max;
    min_ = min;
}

void ColorbarPlugin::init(igl::opengl::glfw::Viewer* _viewer)
{
    igl::opengl::glfw::ViewerPlugin::init(_viewer);

    if (_viewer)
    {
        if (!context_)
        {
            static ImGuiContext* __global_context = ImGui::CreateContext();
            context_ = __global_context;
        }

        const char* glsl_version = "#version 150";
        ImGui_ImplGlfw_InitForOpenGL(viewer->window, false);
        ImGui_ImplOpenGL3_Init(glsl_version);
        ImGui::StyleColorsDark();

        init_colormaps();
    }
}

bool ColorbarPlugin::pre_draw()
{
    glfwPollEvents();

    ImGui_ImplOpenGL3_NewFrame();
    ImGui_ImplGlfw_NewFrame();
    ImGui::NewFrame();
    return false;
}


bool ColorbarPlugin::post_draw()
{
    Eigen::Vector4f bgcolor(0, 0, 0.04, 1);
    draw_colorbar(igl::COLOR_MAP_TYPE_JET, max_, min_ ,bgcolor);

    ImGui::Render();
    ImGui_ImplOpenGL3_RenderDrawData(ImGui::GetDrawData());
    return FALSE;
}

void ColorbarPlugin::shutdown()
{
    // Cleanup
    ImGui_ImplOpenGL3_Shutdown();
    ImGui_ImplGlfw_Shutdown();
}

In my main function:

  // Plot Colorbar
  ColorbarPlugin cbar;
  auto min_max = std::minmax_element(std::begin(roughness_), std::end(roughness_));
  cbar.set_max_min(*min_max.first, *min_max.second);
  viewer.plugins.push_back((igl::opengl::glfw::ViewerPlugin*)&cbar);

  viewer.launch();

Then I got the assert break: 
    IM_ASSERT((g.FrameCount == 0 || g.FrameCountEnded == g.FrameCount)  && "Forgot to call Render() or EndFrame() at the end of the previous frame?");
This is triggered by ImGui::NewFrame(); in ColorbarPlugin::pre_draw()
I don't know how to solve this issue. Probably need to spend some time on ImGUI.

**Comment by lslog3** [comment]
> No

I am able to display the colorbar now. But why the colorbar display a single color rather than  gradually changing colors?
![colorbar](https://user-images.githubusercontent.com/33672316/84605307-3ec96400-ae94-11ea-84ce-86c574bdf2e8.JPG)
